pip/pip-282.md

```diff
@@ -0,0 +1,283 @@
+# Background knowledge
```

Comment:
Can this design avoid the following scenarios? If so, describe how to avoid it.

**context**
- there has 2 consumer: `c1`, `c2`
- there has 3 messages in the backlog: `m1(k1)`, `m2(k2)`, `m3(k3)`

**flow-1**
| time | `add consumer c3` | `delivery messages to client` |
| --- | --- | --- |
| 1 | add consumer to the selector |
| 2 | | route messages: `m1(k1) -> c1`, `m2(k2) -> c2`, `m3(k3) -> c3`
| 3 | | the consumer `c3` was not included in the collection `recentJoinedConsumers`, so move `m3` into the Relay Queue | 
| 4 | mark the recent join position `m3` for the consumer `c3` |

Then the delivery of messages would be stuck.

Reply:
This proposal doesn't avoid your scenario.

It seems your scenario was introduced from https://github.com/apache/pulsar/pull/20026 , right?
https://github.com/apache/pulsar/commit/08b28f5c5cbbdc63467455c40106173475ff01e5#diff-c48d5c94842ac8c9a0c9314b207298069f38c8dcfeda4a9886fb3bb1f77843f2

I noticed the issue referred to in this proposal a long time ago. At least before #20026 was introduced.